extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaPartitionAllocator.java

```diff
@@ -0,0 +1,356 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.imps.CuratorFrameworkState;
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+import org.apache.curator.utils.EnsurePath;
+import org.apache.zookeeper.KeeperException;
+
+import org.springframework.beans.factory.BeanInitializationException;
+import org.springframework.beans.factory.FactoryBean;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.event.ContextClosedEvent;
+import org.springframework.integration.kafka.core.BrokerAddress;
+import org.springframework.integration.kafka.core.ConnectionFactory;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+import org.springframework.xd.dirt.core.DeploymentUnitStatus;
+import org.springframework.xd.dirt.zookeeper.Paths;
+import org.springframework.xd.dirt.zookeeper.ZooKeeperUtils;
+
+/**
+ * Is responsible for managing the partition allocation between multiple instances of a Kafka source module
+ * deployed in a stream.
+ *
+ * As a {@link FactoryBean} it will return the partitions that the current module instance should listen to.
+ * The list of partitions is stored in ZooKeeper, and is created when the module is started for the first time.
+ * If multiple module instances are started at the same time, they synchronize via ZooKeeper, and only the first
+ * will end up creating the module list, while the other instances will read it.
+ *
+ * Restarted modules will read the list of partitions that corresponds to their own sequence. Zero-count Kafka source
+ * modules are not supported.
+ *
+ * As an {@link ApplicationListener}
+ *
+ *
+ * @author Marius Bogoevici
+ */
+public class KafkaPartitionAllocator implements InitializingBean, FactoryBean<Partition[]>,
+		ApplicationListener<ContextClosedEvent> {
+
+	private static final Log log = LogFactory.getLog(KafkaPartitionAllocator.class);
+
+	private final String topic;
+
+	private final String partitionList;
+
+	private final ConnectionFactory connectionFactory;
+
+	private final int count;
+
+	private final int sequence;
+
+	private final CuratorFramework client;
+
+	private final SpringXdOffsetManager offsetManager;
+
+	private String moduleName;
+
+	private String streamName;
+
+	private ObjectMapper objectMapper = new ObjectMapper();
+
+	private InterProcessMutex partitionDataMutex;
+
+	private String partitionDataPath;
+
+	private Partition[] partitions;
+
+	public KafkaPartitionAllocator(CuratorFramework client, ConnectionFactory connectionFactory,
```

Comment:
Right; data cannot be read from ZK in the ctor. However I don't think that `moduleName` and `streamName` are read from ZK - or am I mistaken?